### PR TITLE
android jni shim and vpn api for mobile

### PR DESCRIPTION
**Pull request by majestrate**
`make android` target should yield some `apk`s

### CODE REVIEW

File: llarp/handlers/tun.cpp [hunk 1]
```diff
@@ -585,80 +584,120 @@ namespace llarp
...
+        llarp::LogInfo(Name(), " waiting for vpn to start");
+        vpn   = m_LazyVPNPromise.get_future().get();
+        vpnif = vpn.io;
+        if(vpnif == nullptr)
+        {
+          llarp::LogError(Name(), " failed to recieve vpn interface");
+          return false;
+        }
+        llarp::LogInfo(Name(), " got vpn interface");
+        auto self = shared_from_this();
```
Comment (notlesh) [inline llarp/handlers/tun.cpp]: Why `shared_from_this()` instead of just capturing `this`?

Comment (majestrate) [reply llarp/handlers/tun.cpp]: would using `this` be safe?

Comment (notlesh) [reply llarp/handlers/tun.cpp]: I didn't dig too deep there initially, but it looks like `shared_from_this()` has the effect of tying the lifetime of `this` to the `uv_check_t` that gets injected into `libuv`.

Comment (majestrate) [reply llarp/handlers/tun.cpp]: yes that is why i did it

File: jni/lokinet_jni_common.hpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+  const size_t length = env->GetArrayLength(stringJbytes);
+  jbyte* pBytes       = env->GetByteArrayElements(stringJbytes, NULL);
+
+  T result = visit(llarp::string_view((const char*)pBytes, length));
+
+  env->ReleaseByteArrayElements(stringJbytes, pBytes, JNI_ABORT);
+  env->DeleteLocalRef(stringJbytes);
+
+  return std::move(result);
```
[the remark this answers is not in the document]
Comment (jagerman) [reply jni/lokinet_jni_common.hpp]: Yes, and in fact returning an rvalue reference like this is now a warning in some recent compiler versions.

**Pull request merged**